apps/io_tester/io_tester.cc

```diff
@@ -742,6 +753,76 @@ class unlink_class_data : public class_data {
     }
 
 private:
+    bool blocks_discard_mode() const {
+        return _config.shard_info.unlink_type == unlink_type::discard_blocks;
+    }
+
+    future<size_t> issue_ordinary_unlink() {
+        const auto fname = get_filename(_file_id_to_remove);
+        ++_file_id_to_remove;
+
+        return remove_file(fname).then([]{
+            return make_ready_future<size_t>(0u);
+        });
+    }
+
+    future<size_t> issue_blocks_discarding_unlink() {
+        return ensure_file_to_discard_open().then([this]() {
+            return discard_next_block();
+        }).then([this]() {
+            return close_file_if_all_blocks_discarded();
+        }).then([]() {
+            return make_ready_future<size_t>(0u);
+        });
+    }
```

Comment:
Sadly, the behavior of these lines of code differs from the behavior of `fallocate --punch-hole` tool.

When the workload is inspected via `blktrace`, the trim requests are issued only after the flush completes via `xfsaild`. 

I also tested calling unlink() when all blocks are discarded. No difference.

When I added a call to `fsync()` after each discard is issued, then I saw trim requests issued with certain frequency.